First version of DeepMET

steggema opened the pull request:
#### PR description:

This PR introduces a producer for DeepMET, a deep-learning-based missing pT estimator. The producer creates a new MET collection, and a test configuration is included. The plan is to create a separate PR for possible inclusion in central sequences, e.g. for the upcoming ReMiniAOD campaign.

The tensorflow models for this PR are proposed in a separate PR to cms-data/RecoMET-METPUSubtraction https://github.com/cms-data/RecoMET-METPUSubtraction/pull/5

There are different trainings for different years/conditions (2016, 2018, phase 2), and for 2018 and phase 2 also non-response-corrected trainings. 

Presentations:
https://indico.cern.ch/event/912067/contributions/3835851/ (most recent update)
https://indico.cern.ch/event/883809/contributions/3733818/ (CMS week JetMET meeting)
https://indico.cern.ch/event/854654/contributions/3594579/ (first presentation in MET meeting)

Note that an alternative implementation would be to store additional weights for each PFCandidate and calculate MET and possibly jets in a subsequent step. However, we prefer to leave this option to future studies given that we have not checked the performance on jets (and assume some non-trivial effects) and that this would lead to an increase in complexity of the integration and the additional storage required, in particular if we want to have different METs (e.g. response and non-response-corrected) for a single campaigns.

#### PR validation:

The code has been validated by running it in large-scale checks with simulated and data events to evaluate the performance of the algorithm. A test configuration is included. We have not run any memory or timing tests but we suspect that it runs fast and consumes little memory given the models are small compared to most other tensorflow models.

CPU and memory reports can be found under the following links, obtained on 1000 events from the 136.8311_RunJetHT2017F workflow:
https://steggema.web.cern.ch/steggema/cgi-bin/igprof-navigator/deepmet_cpu_reminiaod
https://steggema.web.cern.ch/steggema/cgi-bin/igprof-navigator/deepmet_mem_reminiaod (note that DeepMET does not seem to appear here, supposedly because it's not in the top 1000, see the text file linked below)

Text dumps are also available here: /afs/cern.ch/user/s/steggema/public/DeepMETIntegration/

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

-

@intrepid42 @yongbinfeng

steggema committed:
First version of DeepMET

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29764/15211

- This PR adds an extra 12KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29764/15211/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29764/15211/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

slava77 commented:
please provide some estimates of CPU time and memory use.

Other than just enabling the code in mini/nanoAOD, what else needs to happen (more developments, retraining)?
If nothing else, I think it would be better to add this now at least to the miniAOD step.

steggema commented:
> please provide some estimates of CPU time and memory use.

Will be added.

> 
> Other than just enabling the code in mini/nanoAOD, what else needs to happen (more developments, retraining)?
> If nothing else, I think it would be better to add this now at least to the miniAOD step.

We don't need and foresee any additional developments or training at the moment (except for uncertainty estimates, which are however a completely different matter both content- and code-wise), so we'll add this to the MiniAOD step if you prefer to have this included in this PR.

Given other commitments, I expect an update by sometime next week.

slava77 commented:
> Given other commitments, I expect an update by sometime next week.

OK
Please note

> Code check has found code style and quality issues which could be resolved by applying following patch(s)

the PR tests can not proceed without this addressed

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/RecoMET-METPUSubtraction#5

steggema committed:
Code format

cmsbuild commented:
+1
Tested at: fc76ca57d0055f1b8e81884bcea76e1b5b59327f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d3fcda/6220/summary.html
CMSSW: CMSSW_11_1_X_2020-05-11-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29764/15282

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @steggema (Jan Steggemann) for master.

It involves the following packages:

RecoMET/METPUSubtraction


@perrotta, @cmsbuild, @slava77 can you please review it and eventually sign? Thanks.
@mmarionncern, @gouskos, @ahinzmann, @jdamgov, @jdolen, @nhanvtran, @gkasieczka, @clelange, @schoef, @mariadalfonso, @seemasharmafnal this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
test parameters
- pull_request = cms-data/RecoMET-METPUSubtraction#5

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/RecoMET-METPUSubtraction#5

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-d3fcda/6220/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2697527
 - DQMHistoTests: Total failures: 30
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2697178
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

cmsbuild commented:
+1
Tested at: d1e3bd5095c8741e883cc9b19daa703dac602f5d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f1bf30/6223/summary.html
CMSSW: CMSSW_11_1_X_2020-05-11-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f1bf30/6223/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2697527
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2697206
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

steggema committed:
Add two DeepMET variants to MiniAOD sequences and event content

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29764/15317

- This PR adds an extra 28KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py  modified in PR(s): #29756

cmsbuild commented:
Pull request #29764 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/RecoMET-METPUSubtraction#5

cmsbuild commented:
+1
Tested at: 6264b296602396b14b9dcfa0545f6689e18b21a9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f1bf30/6252/summary.html
CMSSW: CMSSW_11_1_X_2020-05-12-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f1bf30/6252/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 2 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2697527
 - DQMHistoTests: Total failures: 40
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2697168
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 147 log files, 16 edm output root files, 34 DQM output files

steggema commented:
I've updated the PR description with CPU and memory measurements. Please let me know if I should run on other samples or different tests. A copy of the text added to the PR description follows: 

CPU and memory reports can be found under the following links, obtained on 1000 events from the 136.8311_RunJetHT2017F workflow:
https://steggema.web.cern.ch/steggema/cgi-bin/igprof-navigator/deepmet_cpu_reminiaod
https://steggema.web.cern.ch/steggema/cgi-bin/igprof-navigator/deepmet_mem_reminiaod (note that DeepMET does not seem to appear here, supposedly because it's not in the top 1000, see the text file linked below)

Text dumps are also available here: /afs/cern.ch/user/s/steggema/public/DeepMETIntegration/

santocch commented:
+1

slava77 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -452,6 +452,30 @@ def miniAOD_customizeCommon(process):
+    from Configuration.Eras.Modifier_phase2_common_cff import phase2_common
+    phase2_common.toModify(
+        process.slimmedMETsDeep,
+        max_n_pf=cms.uint32(12500),
+        graph_path=cms.string("RecoMET/METPUSubtraction/data/deepmet/deepmet_v1_phase2.pb")
```
please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.

```suggestion
        max_n_pf=12500,
        graph_path="RecoMET/METPUSubtraction/data/deepmet/deepmet_v1_phase2.pb"
```

same for the rest of this file

steggema replied:
Done.

slava77 commented on RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc:
```diff
@@ -0,0 +1,162 @@
+
+namespace {
+  float divide_and_rm_outlier(float val, float norm) {
+    float ret_val = val / norm;
```
it looks like ```norm``` is the same for all calls.
Better compute an inverse of it and do a product here, which is faster. 
Perhaps "scale" will be more intuitive in that case.

steggema replied:
Done (using scale and renaming the function)

slava77 commented on RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc:
```diff
@@ -0,0 +1,162 @@
+  charge_embedding_ = {{-1, 0}, {0, 1}, {1, 2}};
+  pdg_id_embedding_ = {
+      {-211, 0}, {-13, 1}, {-11, 2}, {0, 3}, {1, 4}, {2, 5}, {11, 6}, {13, 7}, {22, 8}, {130, 9}, {211, 10}};
+  ;
```
what is this ```;``` for ?

steggema replied:
Removed.

slava77 commented on RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc:
```diff
@@ -0,0 +1,162 @@
+
+private:
+  const edm::EDGetTokenT<std::vector<pat::PackedCandidate> > pf_token_;
+  float norm_;
+  bool ignore_leptons_;
+  unsigned int max_n_pf_;
```
can be const

steggema replied:
Made const.

slava77 commented on RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc:
```diff
@@ -0,0 +1,162 @@
+
+  tensorflow::Session* session_;
+
+  std::unordered_map<int, int32_t> charge_embedding_;
+  std::unordered_map<int, int32_t> pdg_id_embedding_;
```
these can be static const

steggema replied:
Made static const and moved to inline initialisation.

slava77 commented on RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc:
```diff
@@ -0,0 +1,162 @@
+      norm_(cfg.getParameter<double>("norm_factor")),
+      ignore_leptons_(cfg.getParameter<bool>("ignore_leptons")),
+      max_n_pf_(cfg.getParameter<unsigned int>("max_n_pf")) {
+  session_ = tensorflow::createSession(cache->graph_def);
```
can be moved to the initializer list (before ```{```)

steggema replied:
Done.

slava77 commented on RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc:
```diff
@@ -0,0 +1,162 @@
+  tensorflow::Tensor output(tensorflow::DT_FLOAT, out_shape);
+
+  tensorflow::NamedTensorList input_list = {
+      {"input", input}, {"input_cat0", input_cat0}, {"input_cat1", input_cat1}, {"input_cat2", input_cat2}};
```
static const?
Please check other values if can be made const, if not constexpr.

steggema replied:
Couldn't be static const because the tensors were defined and constructed in the local scope. I've now moved the tensors to the class level, with the advantage that this avoids reconstructing them in each event but will keep memory occupied (so I'm actually not sure it if's better or not).

slava77 replied:
ah, sorry, I was looking apparently at it and seeing only the format definition and missed the obvious second argument in each case, which are per-event.
Local defs should be fine as well as data members.

slava77 commented on RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc:
```diff
@@ -0,0 +1,162 @@
+  edm::Handle<std::vector<pat::PackedCandidate> > pf_h;
+  event.getByToken(pf_token_, pf_h);
+
+  // PF keys [b'PF_dxy', b'PF_dz', b'PF_eta', b'PF_mass', b'PF_pt', b'PF_puppiWeight', b'PF_px', b'PF_py']
```
perhaps move this to the place where the 8 values are filled. It's more directly useful there

slava77 commented on RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc:
```diff
@@ -0,0 +1,162 @@
+}
+
+void DeepMETProducer::produce(edm::Event& event, const edm::EventSetup& setup) {
+  edm::Handle<std::vector<pat::PackedCandidate> > pf_h;
+  event.getByToken(pf_token_, pf_h);
```
```suggestion
  auto const& pfs = event.get(pf_token_);
```
is shorter

steggema replied:
Done.

slava77 commented on RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc:
```diff
@@ -0,0 +1,162 @@
+  py -= py_leptons;
+
+  auto pf_mets = std::make_unique<pat::METCollection>();
+  reco::LeafCandidate::LorentzVector p4(px, py, 0., std::sqrt(px * px + py * py));
```
```suggestion
  reco::Candidate::LorentzVector p4(px, py, 0., std::hypot(px, py));
```

steggema replied:
Done.

slava77 commented on RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc:
```diff
@@ -0,0 +1,162 @@
+
+  auto pf_mets = std::make_unique<pat::METCollection>();
+  reco::LeafCandidate::LorentzVector p4(px, py, 0., std::sqrt(px * px + py * py));
+  const reco::Candidate::Point vtx(0.0, 0.0, 0.0);
+  pf_mets->emplace_back(reco::MET(p4, vtx));
```
IIRC Point is default-filled with zero, this may even work:
```suggestion
  pf_mets->emplace_back(reco::MET(p4, {}));
```

steggema replied:
Done (supposedly zero-filled according to ROOT doc)

steggema committed:
Address review comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29764/15392

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #29764 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

perrotta commented:
please  test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/RecoMET-METPUSubtraction#5

cmsbuild commented:
+1
Tested at: 180b0adada3a006e31a332b0fb19d214a4ab3646
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f1bf30/6333/summary.html
CMSSW: CMSSW_11_1_X_2020-05-14-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

slava77 commented:
The two kinds of METs are produced in this PR as ```reco::MET``` objects (not as ```pat::MET```). Considering the content (only px and py), ```reco::MET``` is more practical than the pat variant. 
Still, considering that the implementation really does not return anything more than px,py, would it be more practical to insert the DeepMETs  in slimmedMETsPuppi or even slimmedMETs in the same way how Calo and CHS METs are done?
OTOH, it seems like a longer term plan may be to have a fully fledged MET based on particle weights. So, perhaps a more complete object is still more meaningful.

@ahinzmann @lathomas was this discussed?

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f1bf30/6333/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2702162
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2701841
 - DQMHistoTests: Total skipped: 319
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 150 log files, 16 edm output root files, 35 DQM output files

steggema commented:
> The two kinds of METs are produced in this PR as `reco::MET` objects (not as `pat::MET`). Considering the content (only px and py), `reco::MET` is more practical than the pat variant.

Just a quick correction (?): The producer currently yields pat::METs, https://github.com/cms-sw/cmssw/pull/29764/files#diff-63cabc35b8b22ca018a8e84322481671R58, but changing it to reco::MET would of course be a straightforward change. I have no preference and simply picked pat::MET such that the MiniAOD content looks more uniform.

slava77 commented on PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py:
```diff
@@ -20,6 +20,8 @@
         'keep *_slimmedMETsNoHF_*_*',
         'keep *_slimmedMETsPuppi_*_*',
+        'keep *_slimmedMETsDeep_*_*',
+        'keep *_slimmedMETsDeepResp_*_*',
```
all other ```slimmedMET*``` objects have some more subtle structure and come from ```PATMETSlimmer```.
There is nothing similarly "slimmed" in the DeepMET case.
I'm thinking ```deepMETs``` for the former.
The "Resp" apparently stands for response-corrected. Is it the particle-level response or some more general MET response? I'm thinking that a more descriptive (perhaps longer) name would be useful.
Considering that "slimmed" can be omitted here, the longer replacement of "Resp" may not change the total length of the product.

To fit the MET object framework, the "Resp" case may need to enter as a correction to the ```deepMETs``` (that's unless I'm overloading the meaning of "correction" here).

steggema replied:
> all other `slimmedMET*` objects have some more subtle structure and come from `PATMETSlimmer`.
> There is nothing similarly "slimmed" in the DeepMET case.
> I'm thinking `deepMETs` for the former.

Sounds good to me.

> The "Resp" apparently stands for response-corrected. Is it the particle-level response or some more general MET response? I'm thinking that a more descriptive (perhaps longer) name would be useful.
> Considering that "slimmed" can be omitted here, the longer replacement of "Resp" may not change the total length of the product.
> 
> To fit the MET object framework, the "Resp" case may need to enter as a correction to the `deepMETs` (that's unless I'm overloading the meaning of "correction" here).

"Response-corrected" is actually a bit of a misnomer: The two DeepMET versions have a different loss. The first version uses mean-squared-error loss, and has more optimal variance in the bias-variance trade-off and tends to be better for searches in MET resolution tails. The second version has an additional term in the loss yielding a better response and is such better-suited for analyses where unity response matters, in particular at low recoil. There is hence no correction applied. Would something like "deepMETsUnityResp" or "deepMETsUnityResponse" be more clear? (though this would technically only be true beyond some threshold)

slava77 replied:
I see. So, Resp is not really a modifier of the former. Perhaps a more balanced naming is better than.
Still a bit offhand, how about ```deepMETsResponseTune``` and ```deepMETsResolutionTune``` or even  shorter ```deepMETsForResponse``` and ```deepMETsForResolution``` ... or ```deepMETsMinTails``` and ```deepMETsMinBias```?

slava77 commented:
> > The two kinds of METs are produced in this PR as `reco::MET` objects (not as `pat::MET`). Considering the content (only px and py), `reco::MET` is more practical than the pat variant.
> 
> Just a quick correction (?): The producer currently yields pat::METs, https://github.com/cms-sw/cmssw/pull/29764/files#diff-63cabc35b8b22ca018a8e84322481671R58, but changing it to reco::MET would of course be a straightforward change. I have no preference and simply picked pat::MET such that the MiniAOD content looks more uniform.

ah, you are right, I was thinking of ```   pf_mets->emplace_back(reco::MET(p4, {}));```

slava77 commented:
>  Still, considering that the implementation really does not return anything more than px,py, would it be more practical to insert the DeepMETs in slimmedMETsPuppi or even slimmedMETs in the same way how Calo and CHS METs are done?

I had in mind this
https://github.com/cms-sw/cmssw/blob/424ad43856c2c739d702c240187401b7b08566ea/PhysicsTools/PatAlgos/plugins/PATMETSlimmer.cc#L80-L82

steggema committed:
Add DeepMET variants to corrected MET variants, and add them to the regular pat::MET as correction instead of saving them directly in the event

steggema committed:
Make adding DeepMETs in PATMETSlimmer configurable and only activate for regular MiniAOD production so it does not interfere with other usage

cmsbuild commented:
The code-checks are being triggered in jenkins.

steggema commented:
> > Still, considering that the implementation really does not return anything more than px,py, would it be more practical to insert the DeepMETs in slimmedMETsPuppi or even slimmedMETs in the same way how Calo and CHS METs are done?
> 
> I had in mind this
> 
> https://github.com/cms-sw/cmssw/blob/424ad43856c2c739d702c240187401b7b08566ea/PhysicsTools/PatAlgos/plugins/PATMETSlimmer.cc#L80-L82

I've updated using this solution. Note that the PATMETSlimmer is apparently also always re-run in NanoAOD, so I switched adding the DeepMET variants off by default and only add them now in regular MiniAOD processing after a quick exchange with @peruzzim . Adding the DeepMET px and py values to NanoAOD will require a bit of fiddling with the according sequence in Nano.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-29764/15476

- This PR adds an extra 40KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/PatCandidates/src/MET.cc  modified in PR(s): #29863

cmsbuild commented:
Pull request #29764 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.
Tested with other pull request(s) cms-data/RecoMET-METPUSubtraction#5

cmsbuild commented:
+1
Tested at: 74e15dab066a5777e6d1c5f9f0c07dc1930ef3e9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f1bf30/6391/summary.html
CMSSW: CMSSW_11_1_X_2020-05-18-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f1bf30/6391/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 50 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2694466
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2694414
 - DQMHistoTests: Total skipped: 49
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 150 log files, 16 edm output root files, 35 DQM output files

slava77 commented:
> > > Still, considering that the implementation really does not return anything more than px,py, would it be more practical to insert the DeepMETs in slimmedMETsPuppi or even slimmedMETs in the same way how Calo and CHS METs are done?
> > 
> > 
> > I had in mind this
> > https://github.com/cms-sw/cmssw/blob/424ad43856c2c739d702c240187401b7b08566ea/PhysicsTools/PatAlgos/plugins/PATMETSlimmer.cc#L80-L82
> 
> I've updated using this solution. Note that the PATMETSlimmer is apparently also always re-run in NanoAOD, so I switched adding the DeepMET variants off by default and only add them now in regular MiniAOD processing after a quick exchange with @peruzzim . Adding the DeepMET px and py values to NanoAOD will require a bit of fiddling with the according sequence in Nano.

@ahinzmann @lathomas please clarify if this works OK for JME.
Thank you.

ahinzmann commented:
> > > > Still, considering that the implementation really does not return anything more than px,py, would it be more practical to insert the DeepMETs in slimmedMETsPuppi or even slimmedMETs in the same way how Calo and CHS METs are done?
> > > 
> > > 
> > > I had in mind this
> > > https://github.com/cms-sw/cmssw/blob/424ad43856c2c739d702c240187401b7b08566ea/PhysicsTools/PatAlgos/plugins/PATMETSlimmer.cc#L80-L82
> > 
> > 
> > I've updated using this solution. Note that the PATMETSlimmer is apparently also always re-run in NanoAOD, so I switched adding the DeepMET variants off by default and only add them now in regular MiniAOD processing after a quick exchange with @peruzzim . Adding the DeepMET px and py values to NanoAOD will require a bit of fiddling with the according sequence in Nano.
> 
> @ahinzmann @lathomas please clarify if this works OK for JME.
> Thank you.

Yes, that's fine. As long as the fiddling with the sequence for JME-extended-Nano is doable based on this PR for MiniAOD that's fine.

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/29764 74e15da
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show differences only in the miniAOD workflows in the "corrections" of the ```slimmedMETs``` and ```slimmedMETsPuppi``` (similar to the way of storing trk, chs, and calo METs)

silviodonato commented:
merge
@santocch

santocch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will be automatically merged.
